Sources/NIOCore/ByteBuffer-core.swift

```diff
@@ -1119,3 +1117,93 @@ extension ByteBuffer {
         return Range<Int>(uncheckedBounds: (lower: indexFromReaderIndex, upper: upperBound))
     }
 }
+
+// MARK: HexDump helpers
+
+extension ByteBuffer {
+
+    /// Return a `String` of hexadecimal digits of bytes in the Buffer,
+    /// in a format that's compatible with `xxd -r -p`.
+    func hexDump() -> String {
+        return self._storage.dumpBytes(slice: self._storage.fullSlice,
+                                       offset: 0,
+                                       length: self.readableBytes)
+    }
+
+    /// Return a `String` of hexadecimal digits of bytes in the Buffer.,
+    /// in a format that's compatible with `xxd -r -p`, but clips the output to the max length of `maxBytes` bytes.
+    /// If the buffer contains more than the `maxBytes` bytes, this function will return the first `maxBytes/2`
+    /// and the last `maxBytes/2` of that, replacing the rest with `...`, i.e. `01 02 03 ... 09 11 12`.
+    func hexDump(maxBytes: Int) -> String {
+        let totalBytes = self.readableBytes
+
+        if totalBytes <= maxBytes {
+            // If the whole buffer is smaller than the max hexdump length requested, return the full dump.
+            return self.hexDump()
+        } else {
+            // Otherwise, return two concatenated hexdumps of first maxBytes/2 bytes, and last maxBytes/2 bytes.
+            let clipLength = maxBytes / 2
+            let startHex = self._storage.dumpBytes(slice: self._storage.fullSlice,
+                                                   offset: 0,
+                                                   length: clipLength)
+
+            let endHex = self._storage.dumpBytes(slice: self._storage.fullSlice,
+                                                 offset: totalBytes - clipLength,
+                                                 length: clipLength)
+
+            return startHex + " ... " + endHex
+        }
+    }
+
+    /// Returns a `String` of hexadecimal digits of bytes in the Buffer,
+    /// with formatting compatible with output of `hexdump -C`.
+    func hexDumpWithASCII() -> String {
+        var result = ""
+
+        // hexdump -C dumps into lines of upto 16 bytes each
+        // with a last line containing only the offset column.
+        for lineIdx in 0 ... self.readableBytes/16 {
```

Comment:
if you call `get*` later, your indices should be between `self.readerIndex ..< self.writerIndex` not starting at 0.

You can easily generate a test case that will have a higher than 0 readerIndex by doing:

```swift
var buffer = ByteBuffer(string: "XXXXXHello World")
let xxxxx = buffer.readString(length: 5)!
XCTAssertEqual("XXXXX", xxxxx)
XCTAssertEqual(5, buffer.readerIndex)
XCTAssertNil(buffer.getInteger(at: 0, as: UInt8.self)) // <-- nil becasuse 0 is not a readable index now
```

Reply:
@weissi, okay, I think I got the workaround for `readableBytes` — using the same mechanism as the `Storage.dumpBytes`. https://github.com/apple/swift-nio/pull/2475/files#diff-3e3180bab09c17b6054fc18892bb5ee1cd211963a7a2d42265cb905269b98240R1178